### DISCUSSION TITLE
Rebuild widgets after background updates to moor db

### DISCUSSION THREAD

**Discussion by frankvollebregt** [simolus3/drift#1467] [Q&A] [answered]
### intro and context
Hello! I've been experimenting with updating the database contents when the app is in the background. To this end, I use `createMoorIsolate` in my `onBackgroundMessage` callback, fetch and update the data as needed, and end by calling `shutdownAll()` on the `MoorIsolate`. So far, so good.

If the app is inactive (completely closed), this works as expected. The next time the app launches, the new data is there ready to go.

### the problem
When, however, the app is still `paused` in the background, and then `resumed`, the `StreamBuilder` that is `watch`ing the data does not trigger a rebuild (as the data was altered from a different DB accessor, this makes sense).

What I want to do is somehow tell the widget to rebuild either every time I resume the app (to be sure), or somehow trigger a rebuild after updating the data from my background isolate.

If it matters, I'm using `Offstage` widgets for the different tabs of my bottom navigation.

Is there a way to do this? Or should I not do something like this at all? Thanks in advance for the help :)

**Comment by simolus3** [comment, marked as answer]
I think the easiest, but somewhat unclean approach here is to have moor act as if all tables updated after resuming your app. Essentially, you could add this to your database class:

```dart
void refreshAll() => markTablesUpdated(allTables.toSet());
```

You could then call `database.refreshAll()` in response to your app being resumed.

----

A somewhat cleaner, but much more challenging approach would be to entirely synchronize the databases between the UI and the background service (this could also be helpful when the app and the background worker are potentially active at the same time). The isolates can communicate through send/receive ports, and moor can transparently handle that with a [moor isolate](https://moor.simonbinder.eu/docs/advanced-features/isolates/). I am currently looking at ways to make this easier to use.

But if you know that the background worker will only open the database when the app is closed (for instance because it's a FCM background message), the first approach will suffice.

**Reply by frankvollebregt** [reply]
Thank you! The trigger is indeed an FCM background message, so it should work well enough for my use case :)
